File: apps/worker/src/summary/guards.ts
/**
 * Pre-publication guards for generated Russian summaries.
 * Each guard returns { ok: true } or { ok: false, reason: string }.
 */

export interface GuardResult {
  ok: boolean;
  reason?: string;
}

/** Verify the summary body length is within the allowed character range. */
export function guardLength(body: string, min: number, max: number): GuardResult {
  const len = body.length;
  if (len < min) return { ok: false, reason: `too_short:${len}<${min}` };
  if (len > max) return { ok: false, reason: `too_long:${len}>${max}` };
  return { ok: true };
}

const FORBIDDEN_WORDS = ['ужас', 'кошмар', 'шок', 'сенсация', 'скандал века'];

/** Detect forbidden sensational language. */
export function guardForbiddenWords(text: string): GuardResult {
  const lower = text.toLowerCase();
  for (const word of FORBIDDEN_WORDS) {
    if (lower.includes(word)) {
      return { ok: false, reason: `forbidden_word:${word}` };
    }
  }
  return { ok: true };
}

/** Extract all numeric sequences (including those with % or decimal separators). */
function extractNumbers(text: string): string[] {
  return text.match(/\d+(?:[.,]\d+)?%?/g) ?? [];
}

/**
 * Verify that every number found in the source Hebrew titles also appears
 * in the generated Russian text. Numbers must be preserved exactly.
 */
export function guardNumbers(sourceTitles: string[], generatedText: string): GuardResult {
  const sourceNums = new Set(sourceTitles.flatMap(extractNumbers));
  if (sourceNums.size === 0) return { ok: true };

  const genNums = new Set(extractNumbers(generatedText));
  const missing = [...sourceNums].filter(n => !genNums.has(n));
  if (missing.length > 0) {
    return { ok: false, reason: `missing_numbers:${missing.join(',')}` };
  }
  return { ok: true };
}

/**
 * For high-risk stories (casualties, terror, emergencies), the body must
 * contain the attribution phrase "по данным источников".
 */
export function guardHighRisk(body: string, riskLevel: string): GuardResult {
  if (riskLevel !== 'high') return { ok: true };
  if (!body.toLowerCase().includes('по данным источников')) {
    return { ok: false, reason: 'high_risk_requires_attribution' };
  }
  return { ok: true };
}

File: apps/local-engine/summary/guards.py
"""summary/guards.py — Pre-publication guards for generated Russian summaries.

Exact port of apps/worker/src/summary/guards.ts.
Each guard returns a GuardResult(ok, reason).
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field


@dataclass
class GuardResult:
    ok: bool
    reason: str | None = field(default=None)


_FORBIDDEN_WORDS: list[str] = ["ужас", "кошмар", "шок", "сенсация", "скандал века"]

_NUMBER_RE = re.compile(r"\d+(?:[.,]\d+)?%?")


def guard_length(body: str, min_len: int, max_len: int) -> GuardResult:
    """Verify the summary body length is within the allowed character range."""
    n = len(body)
    if n < min_len:
        return GuardResult(ok=False, reason=f"too_short:{n}<{min_len}")
    if n > max_len:
        return GuardResult(ok=False, reason=f"too_long:{n}>{max_len}")
    return GuardResult(ok=True)


def guard_forbidden_words(text: str) -> GuardResult:
    """Detect forbidden sensational language."""
    lower = text.lower()
    for word in _FORBIDDEN_WORDS:
        if word in lower:
            return GuardResult(ok=False, reason=f"forbidden_word:{word}")
    return GuardResult(ok=True)


def _extract_numbers(text: str) -> list[str]:
    return _NUMBER_RE.findall(text)


def guard_numbers(source_titles: list[str], generated_text: str) -> GuardResult:
    """Verify that every number from source Hebrew titles appears in the generated text."""
    source_nums: set[str] = set()
    for title in source_titles:
        source_nums.update(_extract_numbers(title))

    if not source_nums:
        return GuardResult(ok=True)

    gen_nums = set(_extract_numbers(generated_text))
    missing = [n for n in source_nums if n not in gen_nums]
    if missing:
        return GuardResult(ok=False, reason=f"missing_numbers:{','.join(missing)}")
    return GuardResult(ok=True)


def guard_high_risk(body: str, risk_level: str) -> GuardResult:
    """For high-risk stories, body must contain the attribution phrase."""
    if risk_level != "high":
        return GuardResult(ok=True)
    if "по данным источников" not in body.lower():
        return GuardResult(ok=False, reason="high_risk_requires_attribution")
    return GuardResult(ok=True)
